### ISSUE TITLE
Add new phone number login method

### ISSUE THREAD

**Issue by nurtia** [charliewolf/pynder#177]
It is now possible to login via phone number, the facebook authentication is not necessary in this case. It would be great if this would be added, as the facebook authentication is annoying and leads to blocked accounts.

**Comment by derblub** [comment]
I'm guessing this will need some major overhaul, since no more fb-id and -token are needed anymore, right?

**Comment by nurtia** [comment]
Right, I haven't looked into how the session is currently managed but I would assume that with fb-id and token you get a tinder session, and this is passed along the api. And I would hope that now you can just use phone+password to get a tinder session, which you can pass instead. But I have no clue :)

**Comment by GreatOldOne** [comment]
I would gladly check it out. It uses Facebook's Accountkit for SMS verifications, but I assume it actually stores the phone number for phone accounts.
However, the owner of the repository seems to have gone MIA and even a simple timout field isn't getting merged.
I have a heavily altered Pynder at the moment, which includes updates to client 7.4.0, acccount delete method, etc, but again, not much hope in creating pull requests.

As for Facebook blocks, you can prevent them by avoiding logging to retrieve the token and instead using a Proxy Debugger to get the 60-days long token from the Tinder client.

**Comment by nurtia** [comment]
Well, that's the great thing about open source/github: You can create a fork at any time. Would you be up for forking and maintaining the fork? 
I also have a couple of uncommitted fixes, but the new login method would really be worth a fork. 

Regarding blocking: I tried Fiddler, but didn't get it to run. Maybe Charles is easier to set-up, but it's paid -.-

**Comment by charliewolf** [comment]
Hi @GreatOldOne I'm not MIA, I didn't merge your timeout PR because I don't agree with the implementation. As @nurtia said, you're more than welcome to fork. I'm pretty open to merging PRs, if you [look here you'll see quite a few of them](https://github.com/charliewolf/pynder/pulls?q=is%3Apr+is%3Aclosed). Currently the only PRs open are the timeout one, someone adding a deprecated field that Tinder doesn't populate anymore, and three with failing tests.

**Comment by GreatOldOne** [comment]
@nurtia Unfortunately I think I'll be abandoning Tinder soon. I live in a small town and Tinder doesn't seem to give good results for small towns after all the paid modules added to it, so maintaining a fork is out of the question :\.

As for Fiddler, I've never used it, but Charles is only partly payware. For one, it says it has a 30 days trial period but I've been using it for 3 months now. Other than that, you can only have it running for 30 minutes before you need to restart it, but if you only need to get the token, you can do that in less than 5 minutes.

@charliewolf Well, the last merges were the ones I submitted quite a while ago.
The only objection which you made clear for the timout PR was it being an optional argument for Session(), which I explained the why I made it so, and then no more rebuttals from you.

**Comment by brenly** [comment]
@GreatOldOne Fortunately for you, you can set your GPS to the nearest city and budget trips accordingly ;)

**Comment by derblub** [comment]
I'm struggling to capture the SSL requests on my Android device (failed on v7 and v6).
If someone manages to dump the accountkit sms-verification /-signin requests, that would really help a lot.

**Comment by DylanVanAssche** [comment]
@derblub will this help? https://github.com/DylanVanAssche/harbour-sailfinder/blob/master/qml/backend/sailfinder/authenticate.py#L46-L52

**Comment by derblub** [comment]
@DylanVanAssche I've seen and tried those two already, but unfortunately, they don't seem to be valid api-endpoints. At least not anymore.

**Comment by nurtia** [comment]
@derblub SSL capture now failing for me too on v6 (Android N clearly won't work anymore) -- didn't try it in the last month, so looks like they changed something? Is any still able to obtain longterm FB tokens here?

**Comment by nurtia** [comment]
I tried an android emulator with version 5, and different versions of Tinder. I always get the "login failed" error from Tinder. Charles and ssl certificate correctly installed.

**Comment by derblub** [comment]
@nurtia Yes, I've tried that too already.
Also tried to recompile the apk with modified code, but couldn't get the requests to work. Besides ssl-pinning, there seem to be some other stuff implemented, to make reverse-engineering and recompiling more difficult.

Btw: I know all api-endpoints now (easy to get them, from the de-compiled code). What's missing is the request-bodies (and -headers). I couldn't wrap my head around their code, to rebuild the parameters yet.

**Comment by nurtia** [comment]
@derblub Ha nice, that would have been my next step to try. Neat, that you already did that -- and it sucks that also this didn't work. It looks to me as if they did something on the server side. I used an old Tinder version with which I had previously success sniffing the ssl traffic. 

Can you send me the decompiled code (or how did you decompile it?) I wonder if we can do anything at all if we know the account kit api-endpoints, because they likely will require that from our code too?

**Comment by derblub** [comment]
@nurtia To get readable code (instead of just byte-code), you can use [jadx](/skylot/jadx). After intsalling, use this command to start de-compilation:  
`jadx-gui --show-bad-code com.tinder_7.xxx.apk`

**Comment by nurtia** [comment]
@derblub Awesome, thank you very much, that was really helpful!
I'm looking into the documentation of account kit now: https://developers.facebook.com/docs/accountkit/android

In the manifest there is this ClientToken:

`<meta-data android:name="com.facebook.sdk.ApplicationId" android:value="@string/app_id" />
        <meta-data android:name="com.facebook.accountkit.ApplicationName" android:value="@string/app_name" />
        <meta-data android:name="com.facebook.accountkit.ClientToken" android:value="d1891abb4b0bcdfa0580d9b839f4a522" />`

We need to figure out how they do "Check for Existing Sessions", so roughly: One would create an account via the app via SMS. Then we can create an app that gives us the session on the current phone via account kit, and then we use this session for requests. Gosh, why couldn't they just have done username+password *sigh*

According to this: https://developers.facebook.com/docs/accountkit/accesstokens
there are is a Client Access Token Flow that can be enabled or disabled. By default this is enabled, and we get a long-lived access token. I wonder if this long-lived access token is the same as the one we would obtain if we were able to intercept the ssl traffic? When we get this long-lived token, can we then continue using Pynder as before?

Hm, so it looks like the way to get a long lived-token is: request short-lived token (that's something we can do), then the Tinder server requests a long-lived token (we have no chance to do that?), and then the server sends the long-lived token back to the client. Is it at all possible to pretend being that app?

**Comment by derblub** [comment]
All of that should be possible, yes. We just need to re-built the proper request headers and bodies. Hopefully I'll have some more time next week again, to continue reverse-engineering.

**Comment by SokolskyNikita** [comment]
You can now simply capture traffic within your browser, as Tinder has launched a web app: https://tinder.com/app/messages/

**Comment by DylanVanAssche** [comment]
Is the API different on their website? The old one from the app is a mess with things that aren't supported anymore but still send with static data to keep old clients alive.

**Comment by nurtia** [comment]
I didn't manage to intercept the SSL traffic, it requires HTST, Charles root certificate installed and trusted everywhere. Anyone more success?

**Comment by SokolskyNikita** [comment]
@DylanVanAssche it looks simple enough when I check the browser request history on Chrome console.

**Comment by nurtia** [comment]
@SokolskyNikita Is it using a short-lived or long-lived token?

**Comment by DylanVanAssche** [comment]
@SokolskyNikita Will take a look at in a couple months

**Comment by nurtia** [comment]
I tried bypassing the SSL pinning following this guide: https://ritcsec.wordpress.com/2016/12/11/bypassing-certificate-pinning-on-tinder/ -- but now the app just crashes after the FB callback on login.
Did anybody manage to do that successfully?

**Comment by DylanVanAssche** [comment]
@nurtia
It's a long time ago I scraped the API so I don't know how to do it with their new security measurements.
I suggest to take a look at the Web version which seems easier to crack.

**Comment by nurtia** [comment]
@DylanVanAssche It's primarily about the long-lived token.

**Comment by DylanVanAssche** [comment]
@nurtia Oh okay

**Comment by nurtia** [comment]
I'm resolving this issue because I solved the underlying reason for creating it.
We found out that the new phone login method is handled via Facebook AccountKit and thus has the same issues with tokens expiring and automating it eventually will lead to blocked accounts (in that case likely the phone number).
I finally managed to retrieve a long-lived access token by inserting the following code

```
    const-string v9, "fbaccesstoken="
    invoke-static {v9, v0}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
```

in AuthenticationManager.smali below
```
    invoke-virtual {v0}, Lcom/facebook/AccessToken;->getToken()Ljava/lang/String;
    move-result-object v0
```

This simply logs the token which then can be retrieved with logcat. I verified this is indeed a long-lived access token. 

What didn't work: JustTrustMe, Packet Capture, Android SSL TrustKiller, Charles
